deploy/charts/cert-manager/crds/clusterissuers.yaml

```diff
@@ -4,6 +4,7 @@ metadata:
   name: clusterissuers.cert-manager.io
   annotations:
     cert-manager.io/inject-ca-from-secret: cert-manager/cert-manager-webhook-tls
+    helm.sh/hook: crd-install
```

Comment:
This hook causes Helm to not perform *upgrades* on CRD resources after they have been installed, i.e. once installed, they are never modified again.

This is not acceptable for our project, as we make changes to our CRDs in each release.

More info can be found here: https://github.com/helm/helm/pull/6243

Until the Helm team change this and implement actual full lifecycle management for CRDs, there is no way that we can support a change like this.

It's a real shame, but it will absolutely trip users up in future versions and we don't want to go through that same user pain again (v0.4-v0.5 resulted in everyone having all their CRDs *deleted*, and left Helm installations in a broken state due to this annotation).

Reply:
Upgrades are disabled by design as mentioned in the PR and reasons given there. CRD upgrades involve deleting the CRD, deleting all the CRs on the cluster then installing a new CRD. Telling people that we don't offer CRD installs as part of a chart is orthogonal to that technical decision. 

Helm has lifecycle management for installing a CRD and deleting one. This chart has disabled those lifecycles. The fact that helm has taken user feedback about CRD updates being a dangerous process is a good thing. You are in safe hands with the Helm community.

Now what I can not understand is why the community has asked for this feature on multiple occasions and get blocked with comments like this. Every chart must be installable with helm, having out of band processes outside of helm makes for a poor user experience.

Examples
https://github.com/helm/charts/issues/9166
https://github.com/jetstack/cert-manager/issues/1255#issuecomment-513909781
https://github.com/jetstack/cert-manager/pull/1536

I could no doubt find dozens of other comments like this if I spent more than 5mins looking.

Reply:
> Upgrades are disabled by design as mentioned in the PR and reasons given there. CRD upgrades involve deleting the CRD, deleting all the CRs on the cluster then installing a new CRD. Telling people that we don't offer CRD installs as part of a chart is orthogonal to that technical decision.

This is not true. Upgrading a CRD does *not* require deleting all instances of that CRD and re-installing them. This is only true if a publisher/author makes a *backwards incompatible change* to their API schema, which is (naturally) not supported, and is not recommended in any way by #sig-api-machinery.

We have taken a lot of time as a project to set up the necessary infrastructure to support seamless upgrades of CRDs, with the addition of the 'conversion webhook' which will be a part of the v0.14 release (an alpha for which will be available today).

> Helm has lifecycle management for installing a CRD and deleting one

On the contrary, beyond "installing once", Helm *doesn't* have lifecycle management for CRDs. It simply opts to not do anything with them beyond initial install, which is not usable for us as authors of a CRD consumed by many, as part of a project that (naturally) accepts new features and thus, backwards *compatible* API schema changes.

> The fact that helm has taken user feedback about CRD updates being a dangerous process is a good thing.

I don't think I agree 😄 I think there's a lot more Helm could/should do to manage CRDs (notably, just treating them like regular resources).

If the Helm community has come to the conclusion that CRD resources should not be upgraded after first installation, or if they are upgraded, that they are completely uninstalled and re-installed, then I would absolutely like to challenge this status-quo as it is incorrect, and it is tangential to the work #sig-api-machinery has been doing over the years to make CRDs first-class citizens in the Kubernetes apiserver.

Reply:
I've opened https://github.com/helm/helm/issues/7735 in the Helm repository to discuss this further 😄

Reply:
👎  We forked the repo https://hub.helm.sh/charts/infobloxopen/cert-manager

Reply:
🤷‍♂ okay - if you have anything to contribute to the actual Helm issue I've linked above that will allow us to add them to the chart in the actual upstream repo here, please add a comment.

Otherwise, please don't open issues here if something goes wrong after you've gone against our guidance 😅